### ISSUE TITLE
[Windows][Intel x64] Jump table entries seems not valid while targeting MSVC 2022

### ISSUE THREAD

**Issue by amitin** [llvm/llvm-project#95622]
In your project we use LLVM-powered part which compiled with embedded LLVM to an .obj file then gets linked into our main binary. Other parts are in C or C++ and compiled with clang-cl.exe supplied by MSVC 2022 as a ClangCl toolset (at the moment the latest is 17.0.3).
After updating our embedded LLVM from version 16 to 18.1.7 we can see our application crashing. 

Investigations indicate that the crash happened after calculating a jump target from jump table - the resulting jump target value is not correct.
In version 16 a jump table entry was 32-bit, and the result jump target has been calculated by sign-extending jump table entry value, like this:
```
	movabs	r10, offset .LJTI536_0
	movsxd	r9, dword ptr [r10 + 4*r9]
	add	r9, r10
	jmp	r9
```
In version 18 jump table entries are now 64-bit, so the accessing code now doesn't sign-extend entry value
```
	movabs	r10, offset .LJTI536_0
	add	r10, qword ptr [r10 + 8*r9]
	jmp	r10
```

I'm not sure why, but I can see that jump table entry values are still 32-bit. Thus, if values are negative and they don't get sign-extended, then the result jump target becomes invalid. See the screenshot attached.

Is this a bug in LLVM 18? Or some incompatibility issue, for example with MS linker or clang-cl version 17.0.3 which we use for C parts?
Obviously, disabling jump tables fixes the issue for us, but we want them to be enabled.
LMK if you need more info or any other help for this issue.

<img width="942" alt="jump tables" src="https://github.com/llvm/llvm-project/assets/1537180/6e8ffa8f-b5c3-4e79-8bbc-b2663d268592">

**Comment by efriedma-quic** [comment]
If you can reproduce with "llc", that would be helpful.

Are you specifying "large" code model somewhere?  I can reproduce something similar with that.  If that's the case, you're using an unsupported configuration; use "small".  But probably the backend should diagnose this somewhere instead of silently producing nonsense results.

**Comment by amitin** [comment]
@efriedma-quic, yes, we use 'large' code model for Windows intel x64. We cannot use something other than 'large' for intel x64 due to some limitations in our project, but I'll double check. The best way would be to get this supported eventually. :) 
BTW, on Linux x64 it works fine as jump table entries there are the absolute addresses to perform an unconditional indirect jump.
I'll try to create an llc-compilable sample if you still need this.

**Comment by efriedma-quic** [comment]
`llc <  llvm/test/CodeGen/X86/win64-jumptable.ll -mtriple=x86_64-windows-msvc -code-model=large` is enough to reproduce.  The compiler emits `.quad   .LBB0_2-.LJTI0_0`, but the assembler can't actually lower that correctly.

The fact that the assembler incorrectly accepts this appears to be related to c6186922.  CC @MaskRay 

On the compiler side, I suspect 2a2f02e1 is the culprit here; CC @aeubanks @rnk

I'm not sure what you're using "large PIC" for on Windows, though.  PE-COFF can't represent binaries larger than 4GB no matter what code model you use.

**Comment by amitin** [comment]
@efriedma-quic, thank you very much for your investigations, it seems that you've figured it out exactly! Hope that we can get a fix soon, as this issue doesn't appear to be very complicated. Well, at least for my point of view. :)

As for our project, which is in fact a VM for [VAST Platform](url=https://www.instantiations.com/vast-platform/) (Smalltalk), we use 'large' CM as it easier to deal with dynamic code (jit).

**Comment by aeubanks** [comment]
> The compiler emits `.quad .LBB0_2-.LJTI0_0`, but the assembler can't actually lower that correctly.
Why can't the assembler lower this correctly?
> On the compiler side, I suspect [2a2f02e](https://github.com/llvm/llvm-project/commit/2a2f02e19f7737966592f24aa4354286d9756bd9) is the culprit here; CC @aeubanks @rnk
I can undo that patch just for Windows

**Comment by efriedma-quic** [comment]
> > The compiler emits `.quad .LBB0_2-.LJTI0_0`, but the assembler can't actually lower that correctly.
> 
> Why can't the assembler lower this correctly?

The necessary relocation doesn't exist.  The only supported 64-bit relocation is an absolute address.



> > On the compiler side, I suspect [2a2f02e](https://github.com/llvm/llvm-project/commit/2a2f02e19f7737966592f24aa4354286d9756bd9) is the culprit here; CC @aeubanks @rnk
> 
> I can undo that patch just for Windows

Sure.
